[Workers] WASM Rust Guide (#8539)
cloudflare/cloudflare-docs: content/workers/platform/web-assembly/javascript.md

---
title: Wasm in JavaScript
pcx_content_type: javascript-wasm-guide
weight: 1
meta:
  title: Wasm in JavaScript
---

# Invoking Wasm in JavaScript

> Wasm can be used to accelerate existing JavaScript or TypeScript Workers by implementing

Reviewer: ```suggestion
Wasm can be used from within a Worker written in JavaScript or TypeScript by importing a Wasm module, and instantiating an instance of of this module using `[WebAssembly.instantiate()](https://developer.mozilla.org/en-US/docs/WebAssembly/JavaScript_interface/instantiate)`
```

> computationally intensive operations in Wasm, and invoking them from an existing Worker

Reviewer: ```suggestion
In most real-world cases, JavaScript will outperform equivalent Wasm code, because there is a performance cost to loading, instantiating, and calling a Wasm module from JavaScript. But some computationally intensive operations can be faster when executed within a Wasm module.
```

Author: I think this is a strange claim to make, since it would not be true in a pure-Wasm runtime that does not need JS Interop

using the JavaScript WebAssembly API. This guide demonstrates how to invoke a Wasm module
from an existing JavaScript or TypeScript Worker using a simple Wasm binary that
is compiled and used with the JavaScript WebAssembly API, demonstrating the basics of
Wasm and JavaScript interoperability. 

## Simple Wasm Module

> Create a simple Wasm module using the WebAssembly Text Format (`;;` denotes a comment):

Reviewer: @kflansburg - can add a note here about finding Wasm modules or compiling them? Just something to clarify that "how you use Wasm does not mean writing code in a `.wat` file". (know this seems obvious)

Reviewer: e.g. — how can we help people understand that they should never have to deal with this part? Most people this doc is helpful to are probably bringing in some other dependency. Maybe we use one of those to show?

```wat
;; src/simple.wat
(module
  ;; Import a function from JavaScript named `imported_func` 
  ;; which takes a single i32 argument and assign to
  ;; variable $i
  (func $i (import "imports" "imported_func") (param i32))
  ;; Export a function named `exported_func` which takes a
  ;; single i32 argument and returns an i32
  (func (export "exported_func") (param $input i32) (return i32)
    ;; Invoke `imported_func` with $input as argument
    local.get $input
    call $i
    ;; Return $input
    local.get $input
    return
  )
)
```

Using [`wat2wasm`](https://github.com/WebAssembly/wabt), convert the WAT format to WebAssembly Binary Format:

```sh
wat2wasm src/simple.wat -o src/simple.wasm
```

## Bundling

> Wrangler must know to [bundle](/workers/wrangler/bundling/) the Wasm module with your worker so that it can be imported

Reviewer: ```suggestion
Wrangler will [bundle](/workers/wrangler/bundling/) any Wasm module that ends in `.wasm` or `.wasm?module`, so that it is available at runtime within your Worker.
```

in your JavaScript code. Wrangler includes a default bundling rule for Wasm:

```toml

> {"type":"CompiledWasm","globs":["**/*.wasm","**/*.wasm?module"]}

Reviewer: Can omit this

Author: I agree that we should be concise with this, but Wrangler can also seem very magical as a first time user, and I think having some idea of what is happening is useful.

```

If you place `simple.wasm` in a path that matches this glob (such as `src/simple.wasm`), then nothing needs to be done to
configure Wrangler to bundle your module!

> ## Use from JavaScript

Reviewer: Tighter we can be above, and faster we can get to this point, the better. This code example is the part that is most helpful

Next, import and use the Wasm module in your existing JavaScript Worker (TypeScript in this case):

```typescript
import mod from './simple.wasm'

// Define imports available to Wasm instance.
const importObject = {
  imports: {
    imported_func: (arg: number) => {
      console.log(`Hello from JavaScript: ${arg}`)
    }
  },
};

export default {
  async fetch() {
    // Create instance of WebAssembly Module `mod`, supplying
    // the expected imports in `importObject`
    const instance = await WebAssembly.instantiate(mod, importObject);
    // Invoke the `exported_func` from our Wasm Instance with
    // an argument.
    const retval = instance.exports.exported_func(42);
    // Return the return value!
    return new Response(`Success: ${retval}`);
  }
}
```

> When invoked, this Worker should log `Hello from JavaScript: 42` and return `Success: 42`, demonstrating the ability to invoke

Reviewer: Could link to an example project, maybe one that is added to https://github.com/cloudflare/workers-sdk/tree/main/templates, and is [made available at `workers.new/wasm`](https://github.com/cloudflare/workers-sdk/blob/main/packages/workers.new/src/index.ts)?

Wasm methods with arguments from JavaScript and vice versa.

## Next Steps

> In practice, you will likely compile a language of your choice (such as Rust) to WebAssembly binaries. Many languages provide a bindgen to simplify the interaction between JavaScript and Wasm. These tools may integrate with your JavaScript bundler, and provide an API other than

Reviewer: Maybe a way to put this at the top?

the WebAssembly API for initializing and invoking your Wasm module. As an example, see the [documentation](https://rustwasm.github.io/wasm-bindgen/examples/without-a-bundler.html) for Rust's `wasm-bindgen`.

> If you wish to replace your JavaScript worker entirely with Wasm, Workers provides the same rich runtime API when using `workers-rs` in Rust. For more information, see the [Rust guide](/workers/platform/web-assembly/rust/).

Reviewer: Wouldn't say "replace", maybe something that gets at "if this isn't the guide you're looking for, and what you really want is to write a whole worker in Rust..."

---
title: Wasm in JavaScript
pcx_content_type: javascript-wasm-guide
weight: 1
meta:
  title: Wasm in JavaScript
---

# Invoking Wasm from JavaScript

Wasm can be used from within a Worker written in JavaScript or TypeScript by importing a Wasm module, 

> and instantiating an instance of of this module using [`WebAssembly.instantiate()`](https://developer.mozilla.org/en-US/docs/WebAssembly/JavaScript_interface/instantiate). This can be used to accelerate computationally intensive operations which do not involve significant I/O.

Author: ```suggestion
and instantiating an instance of this module using [`WebAssembly.instantiate()`](https://developer.mozilla.org/en-US/docs/WebAssembly/JavaScript_interface/instantiate). This can be used to accelerate computationally intensive operations which do not involve significant I/O.
```

  ;; variable $i
  (func $i (import "imports" "imported_func") (param i32))
  ;; Export a function named `exported_func` which takes a
  ;; single i32 argument and returns an i32
  (func (export "exported_func") (param $input i32) (return i32)
    ;; Invoke `imported_func` with $input as argument
    local.get $input
    call $i
    ;; Return $input
    local.get $input
    return
  )
)
```

Using [`wat2wasm`](https://github.com/WebAssembly/wabt), convert the WAT format to WebAssembly Binary Format:

```sh
wat2wasm src/simple.wat -o src/simple.wasm
```

## Bundling

Wrangler will bundle any Wasm module that ends in `.wasm` or `.wasm?module`, so that it is available at runtime within your Worker. This is done using a default bundling rule which [can be customized](/workers/wrangler/bundling/) in `wrangler.toml`.

## Use from JavaScript

Next, import and use the Wasm module in your existing JavaScript or TypeScript Worker:

```typescript
import mod from './simple.wasm'

// Define imports available to Wasm instance.
const importObject = {
  imports: {
    imported_func: (arg: number) => {
      console.log(`Hello from JavaScript: ${arg}`)
    }
  },

> };
> 
> export default {
>   async fetch() {
>     // Create instance of WebAssembly Module `mod`, supplying
>     // the expected imports in `importObject`
>     const instance = await WebAssembly.instantiate(mod, importObject);

Author: ```suggestion
};

// Create instance of WebAssembly Module `mod`, supplying
// the expected imports in `importObject`. This should be
// done at the top level of the script to avoid instantiation on every request.
const instance = await WebAssembly.instantiate(mod, importObject);
    
export default {
  async fetch() {
```

---
title: Wasm in JavaScript
pcx_content_type: javascript-wasm-guide
weight: 1
meta:
  title: Wasm in JavaScript
---

# Invoking Wasm from JavaScript

Wasm can be used from within a Worker written in JavaScript or TypeScript by importing a Wasm module, 
and instantiating an instance of this module using [`WebAssembly.instantiate()`](https://developer.mozilla.org/en-US/docs/WebAssembly/JavaScript_interface/instantiate). This can be used to accelerate computationally intensive operations which do not involve significant I/O.

This guide demonstrates the basics of Wasm and JavaScript interoperability.

## Simple Wasm Module

In this guide, we will use the WebAssembly Text Format to create a simple Wasm module so that you can
understand how imports and exports work under the hood. In practice, you would not write code in this format — you would instead use the programming language of your choice, and compile directly to WebAssembly Binary Format (`.wasm`).

> Here is our example module (`;;` denotes a comment):

Author: They are creating a file called `src/simple.wat` (as noted in the comment below), which they then convert in the next step.
